transport.go

```diff
@@ -14,221 +13,196 @@ type UnwrapErrorFunc func(nxt DecodeNext) (error, error)
-	// might try to send logs over this transport, which would take the mutex
-	// again. We *must not* call this while we hold the lock. (Yes, we figured
-	// this out by deadlocking ourselves :p)
+func (t *transport) run2() (err error) {
+	// Initialize transport loops
+	readerDone := runInBg(t.readerLoop)
+	writerDone := runInBg(t.writerLoop)
+
+	// Packetize: do work
+	packetizer := newPacketizer(t.dispatcher, t)
+	err = packetizer.Packetize()
+
+	// Log packetizer completion and terminate transport loops
 	t.log.TransportError(err)
+	close(t.stopCh)
+
+	// Wait for loops to finish before resetting
+	<-readerDone
+	<-writerDone
+	t.reset()
 	return
 }
 
-func (t *Transport) Run(bg bool) (err error) {
-	dostart := false
-	t.mutex.Lock()
-	if t.cpkg == nil {
-		err = DisconnectedError{}
-	} else if !t.running {
-		dostart = true
-		t.running = true
-	}
-	t.mutex.Unlock()
-	if dostart {
-		if bg {
-			go t.run2()
-		} else {
-			err = t.run2()
+func (t *transport) readerLoop() {
+	for {
+		select {
+		case <-t.stopCh:
+			return
+		case i := <-t.decodeCh:
+			err := t.cdec.Decode(i)
+			t.readerResultCh <- err
+		case <-t.readByteCh:
+			b, err := t.cdec.ReadByte()
+			res := byteResult{
+				b:   b,
+				err: err,
+			}
+			t.readerResultCh <- res
 		}
 	}
-	return
 }
 
-func (t *Transport) Encode(i interface{}) (err error) {
-	t.wrlck.Lock()
-	defer t.wrlck.Unlock()
-
-	var v1, v2 []byte
-	if v2, err = t.encodeToBytes(i); err != nil {
-		return
-	}
-	l := len(v2)
-	if v1, err = t.encodeToBytes(l); err != nil {
-		return
-	}
-	if err = t.RawWrite(v1); err != nil {
-		return
+func (t *transport) writerLoop() {
+	for {
+		select {
+		case <-t.stopCh:
+			return
+		case bytes := <-t.writeCh:
+			_, err := t.cdec.Write(bytes)
+			t.writerResultCh <- err
+		}
 	}
-	return t.RawWrite(v2)
 }
 
-func (t *Transport) getConPackage() (ret *conPackage, err error) {
-	t.mutex.Lock()
-	ret = t.cpkg
-	t.mutex.Unlock()
-	if ret == nil {
-		err = DisconnectedError{}
-	}
-	return
+func (t *transport) reset() {
+	t.dispatcher.Reset()
+	t.dispatcher = nil
+	t.cdec.Close()
+	t.cdec = nil
 }
 
-func (t *Transport) ReadByte() (b byte, err error) {
-	var cp *conPackage
-	if cp, err = t.getConPackage(); err == nil {
-		b, err = cp.ReadByte()
-	}
-	return
+type byteResult struct {
+	b   byte
+	err error
 }
 
-func (t *Transport) Decode(i interface{}) (err error) {
-	var cp *conPackage
-	if cp, err = t.getConPackage(); err == nil {
-		err = cp.Decode(i)
+func (t *transport) Encode(i interface{}) error {
+	bytes, err := t.EncodeToBytes(i)
+	if err != nil {
+		return err
 	}
+	t.writeCh <- bytes
+	err = <-t.writerResultCh
 	return err
 }
 
-func (t *Transport) RawWrite(b []byte) (err error) {
-	var cp *conPackage
-	if cp, err = t.getConPackage(); err == nil {
-		err = cp.Write(b)
-	}
+func (t *transport) ReadByte() (byte, error) {
+	t.readByteCh <- struct{}{}
+	res := <-t.readerResultCh
+	byteRes, _ := res.(byteResult)
```

Comment:
To circle back to my earlier comment, the reason why I prefer splitting readerResultCh by function is that it would be more consistent -- you're already splitting the input channel by function. I realize it's still correct, but of course you'd have to read the code to ascertain that, instead of having it be guaranteed by the type system.